[web-animations] update accelerated animations for a RenderLayerBacking

graouts opened the pull request:
#### 3311ab9ea59bc6e0016d64581faaf93ff7907ec7
<pre>
[web-animations] update accelerated animations for a RenderLayerBacking
<a href="https://bugs.webkit.org/show_bug.cgi?id=253242">https://bugs.webkit.org/show_bug.cgi?id=253242</a>

Reviewed by Dean Jackson.

Go through the keyframe effect stack associated with a RenderLayerBacking and convert
those effects to accelerated effects provided they can be accelerated. In case we only
find effects that aren&apos;t marked as &quot;running&quot;, which means they won&apos;t produce an interpolated
value that changes over time, we do *not* yield accelerated effects. This means that suspended
or paused animations will not be accelerated unless there are other effects in the stack that
rely on them to produce an interpolated value.

To support this, we change the implementations of KeyframeEffect::canBeAccelerated() and
KeyframeEffect::isRunningAccelerated() when threaded animation resolution is enabled. In this
scenario there are fewer restrictions on the type of animations that can be accelerated, and
we also enforce the &quot;running&quot; state of the associated WebAnimation.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::isRunningAccelerated const):
(WebCore::KeyframeEffect::canBeAccelerated const):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):
* Source/WebCore/rendering/RenderLayerBacking.h:

Canonical link: <a href="https://commits.webkit.org/261120@main">https://commits.webkit.org/261120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2216e96aa32e6e2c66a651bbed408cc7bb555334

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110683 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19764 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2021 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114651 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10876 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/1544 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116424 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102983 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/44094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/12948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18321 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/14886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4199 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -1270,6 +1270,20 @@ void KeyframeEffect::apply(RenderStyle& targetStyle, const Style::ResolutionCont
 
+bool KeyframeEffect::isRunningAccelerated() const
+{
+#if ENABLE(THREADED_ANIMATION_RESOLUTION)
```
is it time to remove this compile guard?

graouts replied:
Not just yet.

graouts committed:
[web-animations] update accelerated animations for a RenderLayerBacking
https://bugs.webkit.org/show_bug.cgi?id=253242

Reviewed by Dean Jackson.

Go through the keyframe effect stack associated with a RenderLayerBacking and convert
those effects to accelerated effects provided they can be accelerated. In case we only
find effects that aren't marked as "running", which means they won't produce an interpolated
value that changes over time, we do *not* yield accelerated effects. This means that suspended
or paused animations will not be accelerated unless there are other effects in the stack that
rely on them to produce an interpolated value.

To support this, we change the implementations of KeyframeEffect::canBeAccelerated() and
KeyframeEffect::isRunningAccelerated() when threaded animation resolution is enabled. In this
scenario there are fewer restrictions on the type of animations that can be accelerated, and
we also enforce the "running" state of the associated WebAnimation.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::isRunningAccelerated const):
(WebCore::KeyframeEffect::canBeAccelerated const):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):
* Source/WebCore/rendering/RenderLayerBacking.h:

Canonical link: https://commits.webkit.org/261120@main
